clientv3/lease.go

```diff
@@ -26,17 +30,254 @@ type (
+	_, ok := l.keepAlives[id]
+	if !ok {
+		l.keepAlives[id] = lc
+		l.deadlines[id] = time.Now()
+		l.mu.Unlock()
+		return lc, nil
+	}
+	l.mu.Unlock()
+
+	resp, err := l.KeepAliveOnce(ctx, id)
+	if err != nil {
+		return nil, err
+	}
+	lc <- resp
+	return lc, nil
+}
+
+func (l *lessor) KeepAliveOnce(ctx context.Context, id lease.LeaseID) (*LeaseKeepAliveResponse, error) {
+	for {
+		resp, err := l.keepAliveOnce(ctx, id)
+		if err == nil {
+			return resp, err
+		}
+
+		nerr := l.switchRemote(err)
+		if nerr != nil {
+			return nil, nerr
+		}
+	}
+}
+
+func (l *lessor) keepAliveOnce(ctx context.Context, id lease.LeaseID) (*LeaseKeepAliveResponse, error) {
+	stream, err := l.getRemote().LeaseKeepAlive(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	err = stream.Send(&pb.LeaseKeepAliveRequest{ID: int64(id)})
+	if err != nil {
+		stream.CloseSend()
+		return nil, err
+	}
+
+	resp, rerr := stream.Recv()
+	stream.CloseSend()
+	if rerr != nil {
+		return nil, rerr
+	}
+	return (*LeaseKeepAliveResponse)(resp), nil
+}
+
+func (l *lessor) recvKeepAliveLoop() {
+	for {
+		stream := l.getKeepAliveStream()
+
+		for {
+			resp, err := stream.Recv()
+			if err != nil {
+				l.switchRemote(err)
+				break
+			}
+
+			l.mu.Lock()
+			lch, ok := l.keepAlives[lease.LeaseID(resp.ID)]
+			if !ok {
+				continue
+			}
+
+			if resp.TTL <= 0 {
+				close(lch)
+				delete(l.deadlines, lease.LeaseID(resp.ID))
+				delete(l.keepAlives, lease.LeaseID(resp.ID))
+			} else {
+				select {
+				case lch <- (*LeaseKeepAliveResponse)(resp):
+					l.deadlines[lease.LeaseID(resp.ID)] =
+						time.Now().Add(1 + time.Duration(resp.TTL/3)*time.Second)
+				default:
+				}
+			}
+			l.mu.Unlock()
+		}
+	}
+}
+
+func (l *lessor) sendKeepAliveLoop() {
+	for {
+		time.Sleep(time.Second / 2)
+		tosend := make([]lease.LeaseID, 0)
+
+		now := time.Now()
+		l.mu.Lock()
+		for id, d := range l.deadlines {
+			if d.Before(now) {
+				tosend = append(tosend, id)
+			}
+		}
+		l.mu.Unlock()
+
+		stream := l.getKeepAliveStream()
+
+		for _, id := range tosend {
+			r := &pb.LeaseKeepAliveRequest{ID: int64(id)}
+			err := stream.Send(r)
+			if err != nil {
+				l.switchRemote(err)
+				break
+			}
+		}
+	}
+}
+
+func (l *lessor) getRemote() pb.LeaseClient {
+	l.mu.Lock()
+	defer l.mu.Unlock()
+	return l.remote
+}
+
+func (l *lessor) switchRemote(prevErr error) error {
+	newConn, err := l.c.retryConnection(l.conn, prevErr)
+	if err != nil {
+		return err
+	}
+
+	l.mu.Lock()
+	defer l.mu.Unlock()
+	if l.stream != nil {
+		l.stream.CloseSend()
+	}
+
+	l.conn = newConn
+	l.remote = pb.NewLeaseClient(l.conn)
+	// todo: pass in context...
+	stream, serr := l.remote.LeaseKeepAlive(context.TODO())
+	// not likely to happen since we just get the connection...
+	if serr != nil {
+		return serr
+	}
+	l.stream = stream
+	return nil
+}
+
+func (l *lessor) getKeepAliveStream() pb.Lease_LeaseKeepAliveClient {
+	l.mu.Lock()
+	defer l.mu.Unlock()
+	if l.stream != nil {
+		return l.stream
+	}
+
+	for {
```

Comment:
we probably should not error this out for the entire life-cycle of the client. the stream should always be here as long as the client exists. we should error this out when the client is closed. make sense? probably we should add a todo?